### PR TITLE
Support buffer modifying linters

### PR DESCRIPTION
**Pull request by steelbrain**
Fixes https://github.com/AtomLinter/Linter/issues/696
I've confirmed that it works.
Buffer modifying linters need to have a `modifiesBuffer:true` in their object. This change is backward compatible.

### CODE REVIEW

File: lib/helpers.coffee [hunk 1]
```diff
@@ -4,6 +4,12 @@ path = require 'path'
 child_process = require 'child_process'
 
 Helpers = module.exports =
+  shouldTriggerLinter: (linter, wasTriggeredOnChange, scopes)->
+    # Trigger fly linters on save, but not save linters on fly
```
Comment (AsaAyers) [inline lib/helpers.coffee]: If you're going to separate these, we need a new category for CoffeeLint which does both.

Comment (steelbrain) [inline lib/helpers.coffee]: Quite confusing, would you please elaborate?

Comment (AsaAyers) [inline lib/helpers.coffee]: Sorry, I misread the comment thinking fly linters would *not* trigger on save. It's only the reverse.

CoffeeLint will lint on the fly, but there are times (often when editing the config file) when I save specifically to trigger the linter.

File: lib/editor-linter.coffee [hunk 2]
```diff
@@ -53,30 +53,34 @@ class EditorLinter
...
+    @triggerLinters(true, wasTriggeredOnChange, scopes).then( =>
+      return Promise.all(@triggerLinters(false, wasTriggeredOnChange, scopes))
+    ).then =>
       @lock(wasTriggeredOnChange, false)
 
   # This method returns an array of promises to be used in lint
-  triggerLinters: (wasTriggeredOnChange, scopes) ->
-    Promises = []
+  triggerLinters: (bufferModifying, wasTriggeredOnChange, scopes) ->
+    ToReturn = if bufferModifying then Promise.resolve() else []
```
Comment (AsaAyers) [inline lib/editor-linter.coffee]: It's not a good idea to have a variable that alternates between different types. Sometimes it's a `Promise`, sometimes it's an `Array`. Off the top of my head I don't know what the exact solution is here, but I don't think you should combine these.

File: lib/editor-linter.coffee [hunk 3]
```diff
@@ -53,30 +53,34 @@ class EditorLinter
...
+          if linter.scope is 'project'
+            @linter.setMessages(linter, results)
+          else
+            # Trigger event instead of updating on purpose, because
+            # we want to make MessageRegistry the central message repo
+            @emitter.emit('should-update', {linter, results})
+        ).catch (error) ->
+          atom.notifications.addError error.message, {detail: error.stack, dismissable: true}
+      if bufferModifying
+        ToReturn.then -> currentLinter()
```
Comment (AsaAyers) [inline lib/editor-linter.coffee]: Do linters that can modify the buffer not get to return messages? This implementation will only keep the messages from the last linter that modified the buffer.

File: lib/editor-linter.coffee [hunk 4]
```diff
@@ -53,30 +53,34 @@ class EditorLinter
...
+      currentLinter = =>
+        return new Promise((resolve) =>
+          resolve(linter.lint(@editor, Helpers))
+        ).then((results) =>
+          if linter.scope is 'project'
+            @linter.setMessages(linter, results)
+          else
+            # Trigger event instead of updating on purpose, because
+            # we want to make MessageRegistry the central message repo
+            @emitter.emit('should-update', {linter, results})
```
Comment (AsaAyers) [inline lib/editor-linter.coffee]: Is `MessageRegistry` in place? Shouldn't we set the value and have it emit to notify the views that they need to update?

I think maybe this is an indirect setter, and I've found that strategy to not work very well as the system grows.

File: lib/editor-linter.coffee [hunk 5]
```diff
@@ -53,30 +53,34 @@ class EditorLinter
...
+      return Promise.all(@triggerLinters(false, wasTriggeredOnChange, scopes))
+    ).then =>
       @lock(wasTriggeredOnChange, false)
 
   # This method returns an array of promises to be used in lint
-  triggerLinters: (wasTriggeredOnChange, scopes) ->
-    Promises = []
+  triggerLinters: (bufferModifying, wasTriggeredOnChange, scopes) ->
+    ToReturn = if bufferModifying then Promise.resolve() else []
     @linter.getLinters().forEach (linter) =>
```
Comment (AsaAyers) [inline lib/editor-linter.coffee]: does `getLinters()` return an Array? Right now this whole `ToReturn` thing is really just simulating a `.map()` or a `.reduce()` depending on if the linters modify the buffer. If you can I recommend `return @linter.getLinters().reduce(...)` instead

**Comment by park9140** [comment]
We really need unit tests before we start making changes like this.

**Comment by park9140** [comment]
Can you explain the full process you are using for handling buffer modifying linters?  I think there are things missing here but I can't really follow your code to point them out in any exact manner.

File: lib/editor-linter.coffee (hunk 3 above)
Comment (steelbrain) [inline lib/editor-linter.coffee]: Not really, it'll keep the messages from all linters, and not only that, It'll only run the bufferModifying linters in series, not in parallel.

File: lib/editor-linter.coffee (hunk 4 above)
Comment (steelbrain) [inline lib/editor-linter.coffee]: While I agree with the approach, I had two options, complexity or this, I chose this. You can find more about how this works in the onShouldUpdate observer in messaeg-registry

File: lib/editor-linter.coffee (hunk 5 above)
Comment (steelbrain) [inline lib/editor-linter.coffee]: `getLinters()` doesn't return an array, It's an ES6 map. I was initially gonna do map/reduce but I figured this approach to be simpler (I tried it first then deleted the code)

**Comment by steelbrain** [comment]
@park9140 

1. Lint method is triggered
2. We trigger all the bufferModifying linters
  1. Non-Buffer modifying linters are skipped
  2. Common logic is stored in a module
  3. a .then handler is added to the `ToReturn` and the return value of common logic function is returned from there
  4. The last promise resolves when all the promises have, in sequence
3. We trigger the non-buffer-modifying linters as usual

File: lib/editor-linter.coffee (hunk 3 above)
Comment (AsaAyers) [inline lib/editor-linter.coffee]: I think you're wrong. A test would prove it one way or the other.

**Comment by steelbrain** [comment]
I think I am gonna add some tests for it

File: lib/editor-linter.coffee (hunk 3 above)
Comment (AsaAyers) [inline lib/editor-linter.coffee]: Oh, I see. It works because you're relying on  side effects. The return value doesn't actually matter. If it did this would fail.

Comment (steelbrain) [inline lib/editor-linter.coffee]: I still wouldn't call it a side-effect though

**Comment by park9140** [comment]
@steelbrain, how does this handle the situation of multiple buffer modifying linters?  Do the buffer modifying linters get run in series?

**Comment by steelbrain** [comment]
They run in sequence, not in parallel, so only one buffer modifying linter would be executed at a time, so it makes it safe to have multiple buffer modifying linters.

**Comment by steelbrain** [comment]
@k2b6s9j your spec about a linter error is logging an error in the console but still failing

**Comment by steelbrain** [comment]
The tests have been pushed people. They are all passing.

File: spec/buffer-modifying-provider-spec.coffee [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        expect(activeEditor.getText()).toBe('second')
+        activeEditor.setText('third')
+        return []
+    normalLinter =
+      grammarScopes: ['*']
+      scope: 'file'
+      lintOnFly: false
+      modifiesBuffer: false
+      lint: ->
+        expect(activeEditor.getText()).toBe('third')
```
Comment (park9140) [inline spec/buffer-modifying-provider-spec.coffee]: expectations should not be part setup, this makes tests hard to follow.  store the information then write the expectations below.

File: spec/buffer-modifying-provider-spec.coffee [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      modifiesBuffer: false
+      lint: ->
+        last = 'normal'
+        return []
+    linter.addLinter(bufferModifying)
+    linter.addLinter(normalLinter)
+    waitsForPromise ->
+      linter.getActiveEditorLinter().lint(false).then ->
+        expect(last).toBe('normal')
+  it 'runs in parallel', ->
```
Comment (park9140) [inline spec/buffer-modifying-provider-spec.coffee]: 'runs buffer modifying linters in series by registration order'?

File: spec/buffer-modifying-provider-spec.coffee (hunk 6 above)
Comment (steelbrain) [inline spec/buffer-modifying-provider-spec.coffee]: it's not really a setup, it demos how first linter set's it's text to first and second linter verifies it and second linter then sets a text and third verifies it. I think that was the best way to test that it runs in a chain.

File: spec/buffer-modifying-provider-spec.coffee (hunk 7 above)
Comment (steelbrain) [inline spec/buffer-modifying-provider-spec.coffee]: already fixed, refresh your page.

File: spec/buffer-modifying-provider-spec.coffee (hunk 6 above)
Comment (park9140) [inline spec/buffer-modifying-provider-spec.coffee]: Everything that does execution of code is setup.

File: spec/buffer-modifying-provider-spec.coffee (hunk 7 above)
Comment (steelbrain) [inline spec/buffer-modifying-provider-spec.coffee]: It was a typo, it should've been `run in sequence`

File: spec/buffer-modifying-provider-spec.coffee (hunk 6 above)
Comment (steelbrain) [inline spec/buffer-modifying-provider-spec.coffee]: :sweat_smile: I don't follow, example please?

File: spec/buffer-modifying-provider-spec.coffee [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+      modifiesBuffer: false
+      lint: ->
+        last = 'normal'
+        return []
+    linter.addLinter(bufferModifying)
+    linter.addLinter(normalLinter)
+    waitsForPromise ->
+      linter.getActiveEditorLinter().lint(false).then ->
+        expect(last).toBe('normal')
+  it 'run in sequence', ->
```
Comment (park9140) [inline spec/buffer-modifying-provider-spec.coffee]: `it runs buffer modifying linters in series`

File: spec/buffer-modifying-provider-spec.coffee (hunk 6 above)
Comment (park9140) [inline spec/buffer-modifying-provider-spec.coffee]: Also this has to many expectations in one test.  You actually have a couple tests going on here that could run off of the same setup.

File: spec/buffer-modifying-provider-spec.coffee (hunk 8 above)
Comment (steelbrain) [inline spec/buffer-modifying-provider-spec.coffee]: we are actually describing `buffer modifying linters` above, so we don't have to write that whole message, when we write `run in sequence` it becomes `buffer modifying linters run in sequence`

File: spec/buffer-modifying-provider-spec.coffee [hunk 9]
```diff
@@ -0,0 +1,77 @@
+describe 'buffer modifying linters', ->
+  getModuleMain = -> atom.packages.getActivePackage('linter').mainModule.instance
+  beforeEach ->
+    waitsForPromise ->
+      atom.packages.activatePackage('linter')
+    waitsForPromise ->
+      atom.workspace.open('test.txt')
+  it 'are triggered before other linters', ->
```
Comment (park9140) [inline spec/buffer-modifying-provider-spec.coffee]: `it 'triggers buffer modifying linters first'`

Comment (steelbrain) [inline spec/buffer-modifying-provider-spec.coffee]: done

File: spec/buffer-modifying-provider-spec.coffee (hunk 6 above)
Comment (steelbrain) [inline spec/buffer-modifying-provider-spec.coffee]: You mean write a function that retuns a provider and then use it in tests instead of writing it each time?

Comment (park9140) [inline spec/buffer-modifying-provider-spec.coffee]: nest a describe for the sequence testing.  Do the setup portions of this then write the individual tests that exist here.
```coffeescript
describe 'order of execution of buffer modifying linters', ->
  beforeEach ->
    // set up all the linters here, copy the editor value when each linter was executed into a variable for that linter
  it 'should execute the linters in order of registration` ->
    expect(secondLinterEditorValue).toBe('first')
  it 'should execute async linters in correct order'
    expect(asyncLinterEditorValue).toBe('second')
  it 'should execute non modifying linters last' ->
    expect(nonModifyingEditorValue).toBe('third')
```

Additionally you need to have timeouts and

Comment (park9140) [inline spec/buffer-modifying-provider-spec.coffee]: More appropriately you could just push the linter onto an array as it is executed then test the position in the array is correct rather than testing the actual editor values...

Comment (steelbrain) [inline spec/buffer-modifying-provider-spec.coffee]: makes sense, thanks!

Comment (steelbrain) [inline spec/buffer-modifying-provider-spec.coffee]: I am a bit confused, we don't really handle the triggering of the linters ourselves, we add them to as linters, linter does the rest.
Also pushing the values to an array and then verifying their indices won't be very different from what's already implemented, do you have an alternative approach in mind?

Comment (park9140) [inline spec/buffer-modifying-provider-spec.coffee]: I mean when you define the linters for the test.  
```
linterExecutionOrder = []
third =
      grammarScopes: ['*']
      scope: 'file'
      lintOnFly: false
      modifiesBuffer: true
      lint: ->
        linterExecutionOrder.push(this)
        activeEditor.setText('third')
        return []

expect(linterExecutionOrder[3]).toBe(third);

Comment (steelbrain) [inline spec/buffer-modifying-provider-spec.coffee]: done

**Comment by steelbrain** [comment]
Any further suggestions?

**Comment by steelbrain** [comment]
Merging in on :+1: from @AsaAyers and because there were no known unresolved bugs.

**Comment by iam4x** [comment]
Woot! We will be able to include https://atom.io/packages/jscs-fixer directly into `linter-jscs` if I'm right!

:clap:

**Comment by steelbrain** [comment]
You are :tada:
